### PR TITLE
FLOPS calculation for matmul operation with 4D inputs in NAS

### PR DESCRIPTION
**Pull request by ljaljushkin**
### Changes

This PR proposes how to support FLOPS calculation for (`matmul`, `bmm`) represented as torch functions on two tensors, not by module with weights. It's not a final solution, subject to discussion and modifications.

Note, there are many limitations: 
- All other operations associated with `matmul`  are not supported: `torch.addmm`, `torch.nn.functional.Linear`, `torch.mm`. 
- covered inputs of the same shape, but matmul also supports: 1D x 2D, 2D x 1D, 1D x ND, ND x 1D https://pytorch.org/docs/stable/generated/torch.matmul.html
- Filter pruning is not covered. Changes were made and tested for NAS part.
- FLOPS calculation was verified for FP32 model only. FLOPS of the pruned model is not verified.
- matmul, which takes learnable parameter on second input, is not supported (GNMT case) 
- `t.matmul(t)` is not supported

### Reason for changes

Currently, NNCF for Torch supports FLOPS calculation for modules only (e.g. `nn.Linear` or `nn.Conv2d`), because all required layer attributes are collected for modules now.
However, modern architectures (e.g., Transformers) are constructed not from modules only. Usually there is a multiplication of two activations represented as `torch.matmul` operation. For such operations, there was no extraction of layer attributes that is needed for FLOPS calculation.

### Related tickets

n/a

### Tests

`test_flops_and_weights_calculation`

### CODE REVIEW

File: nncf/common/graph/layer_attributes.py [hunk 1]
```diff
@@ -125,6 +125,15 @@ def get_target_dim_for_compression(self) -> int:
         return 0
 
 
+class MatmulAttributes(BaseLayerAttributes):
```
Comment (ljaljushkin) [inline nncf/common/graph/layer_attributes.py:128]: Introduced new class for `matmul` specifically, since `LinearLayerAttrbiutes` was designed for modules with weights and biases.

File: nncf/torch/dynamic_graph/wrappers.py [hunk 2]
```diff
@@ -193,6 +214,8 @@ def _execute_op(op_address: 'OperationAddress',
             is_called_inside_nncf_module = isinstance(ctx.get_current_module(), _NNCFModuleMixin)
             node = ctx.maybe_add_node(processed_input, tensor_metas, op_address,
                                       layer_attrs, ignored_algos, is_called_inside_nncf_module)
+            if node is not None:
+                _add_operation_attributes(node, ctx)
```
Comment (ljaljushkin) [inline nncf/torch/dynamic_graph/wrappers.py:218]: Have to override existing attributes for operations that require graph traversal in order to find layer attributes.
input and output features for `matmul` is extracted from the shape of second input.

File: nncf/torch/pruning/utils.py [hunk 3]
```diff
@@ -105,16 +112,21 @@ def collect_output_shapes(graph: NNCFGraph) -> Dict[NNCFNodeName, List[int]]:
...
     modules_out_shapes = {}
+
     output_shape_collecting_info = [
-       (NNCF_GENERAL_CONV_MODULES_DICT, slice(2, None)),
-       (NNCF_LINEAR_MODULES_DICT, slice(None)),
+        OutputShapeCollectionInfo(node_types=[v.op_func_name for v in NNCF_GENERAL_CONV_MODULES_DICT],
+                                  shape_slice=slice(2, None)),
+        OutputShapeCollectionInfo(node_types=[v.op_func_name for v in NNCF_LINEAR_MODULES_DICT],
+                                  shape_slice=slice(None)),
+        OutputShapeCollectionInfo(node_types=['matmul', 'bmm'],
```
Comment (ljaljushkin) [inline nncf/torch/pruning/utils.py:121]: @daniil-lyakhov Does it affect Filter Pruning algorithm?
Does Filter Pruning algorithm calculate FLOPS after setting some meaningful pruning masks, or it only calculates for FP32 model?

Comment (daniil-lyakhov) [reply nncf/torch/pruning/utils.py:121]: FLOPS calculation besides statistics prompt is used to calculate filter pruning level by given FLOPS pruning level. This change possibly affects masks configuration for models that have matmul operations with flops pruning targets in config

File: nncf/torch/dynamic_graph/wrappers.py [hunk 4]
```diff
@@ -169,6 +170,30 @@ def wrapped(self, *args, **kwargs):
...
+    op_address = node.op_exec_context.op_address
+    # TODO(nlyalyus): support other operations: torch.addmm, torch.nn.functional.Linear, torch.bmm, torch.mm
+    if op_address.operator_name in ('matmul', 'bmm'):
+        graph = ctx.graph
+        input_edges = graph.get_input_edges(node)
+        # no input nodes when input_info tensor was bound to a *args or **kwargs on graph tracing
+        if len(input_edges):
+            assert len(input_edges) == 2
+            activation_edge, weight_edge = input_edges
+            activation_shape = activation_edge.activation_shape
```
Comment (daniil-lyakhov) [inline nncf/torch/dynamic_graph/wrappers.py:183]: @ljaljushkin, can we use `args` and `kwargs` for operation attributes? For some operations (like `transpose`) layer attributes collection by input edges is impossible (`input_shape=(5,5,5), transpose=(2, 1, 0)`). You can find example with torch implementation for transpose layer attributes [here](https://github.com/daniil-lyakhov/nncf/commit/079578ccab88a18640ada7088b33f548e63937f4#diff-6e72f489b20539eb2b9d90608c21d9288a65638482687360312d8aebbc546871)
